design/design-doc-stacks.md

```diff
@@ -137,9 +142,6 @@ spec:
  title: Redis stack for Crossplane
  description: "Markdown syntax about how Redis is a really cool database"
  version: 0.1.0
- icons:
- - base64data: iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
```

Comment:
help me understand why this is being removed from the `Extension` record that the SM generates?  this is how the SM is currently storing the image data after extracting it from the stack package.  Why should we remove that?  How else would a client (e.g. a UI) know the icon to display for a given extension record if we remove this?

Reply:
I misunderstood the intent here.  I saw this as duplicative to the OCI stashed icon and assumed that stack icons would be extracted by the consuming tools or websites.

In cluster tools may benefit from having the icon stashed in the CRD.  In that case the external icon file is extraneous.

Is the benefit to including the raw icon file in the image directories a better development cycle?

It may be an even better experience for a stack building tool to accomplish this out of cluster, without extension-manager involvement.  With that thinking we should also suggest\ to processing UI YAML files using external stack tools instead of the extension-manager.

This approach would reduce the unnecessary in-cluster processing, error domains, and scope while also reducing the image size.

The Extension CRD file without the `icons:` stanza could be the one the user edits while a computed version (with `icons` and `ui`) is what is bundled in the stack created by a stack tool.  Users that hand-roll stacks can leave the `icons` and `ui` stanzas alone in the Extension CRD and compute the base64 by hand (likely automating this with `base64` calls from a `Makefile`).

Reply:
I discussed this with @jbw976 and edited my question above to better state my previous line of questioning.

To further clarify, the split here is whether or not the CLI tool, specifically the `build` feature of that tool, should take over some of the responsibilities that the extension-manager may have already assumed or would have to further assume. ( @suskin ) 

My assertions that graphics and UI yaml files would effectively take up twice their bite size has been invalidated.

Since the Extension CRD `icons` stanza would be generated at run-time by the extension-manager there is no duplication of graphic (and potentially UI files) in the package.

While the `icons:` stanza was not intended to be bundled in the package file, only added at run-time by the extension-manager, would there be some advantage to including the `icons:` property attached to some yaml (Extension CRD or otherwise) within the package and keeping the graphic files themselves out of the package?

In either case, the tools will need to identify the nearest icon through the ancestry of (potentially) Extension CRD, resource, group, and app.

I previously made the assumption that the final Kubernetes Extension record may need to include the icon and UI metadata.  However, there may be no clear reason to suspect that in-cluster tools will ever need to portray the UI or graphics.  The UI and graphical portrayal of an extension may exist only to serve the registry index and external package manager tools ( @thephred @rathpc @lukeweber ?)


We are left with a question of processing the UI yaml and graphic files once at developer packaging time (the `stack build` runtime), or every time the package is installed (from every `crossplane` runtime).

A benefit of processing this data at build time is that failure scenarios are reduced, especially on the end-user front.

Reply:
@displague I strongly prefer moving as much logic and complexity to build-time as possible. That way, the runtime logic can be as dumb and fast as possible. When bundles become larger and more complicated, keeping most of the logic on the build-time side means that most of the increase in processing time will happen in the build, which I have learned the hard way to appreciate.

The unbundling and installation logic also becomes easier to understand when the built package won't be mucked around with much at runtime on installation, so that can be nice for users who want to understand what is happening, or for users who need to debug some unexpected results.

Reply:
I think the most strait forward way to deal with this for now is to expect base64 encoded icon files. It seems like we could decide if we want the data to be in the yaml files or if we want to encode directly into the CRD as a separate discussion.

Reply:
Resolving this because the `icon` has been added back to the Extension record.  The separate discussion on encoding into the CRDs is happening here: 
https://github.com/crossplaneio/crossplane/pull/604#discussion_r314188551

We decided against surfacing all of the meta-data files (`resource.yaml`, `group.yaml`, CRD related icons) in the Extension record or the `app.yaml` file.  We decided that it is important that the installed CRDs can be discovered and express their metadata without the parent Extension record.